Ustar and other fixes

alexmohr opened the pull request:
This PR fixes: 
* #10
* #9
* #8
* #7
* #6

alexmohr committed:
ustar: add ustar support

this commit adds support and tests for ustar.

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
ustar: add directory support

this commit adds support for directories when ustar is used

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
general: add platform abstraction

this abstraction allows tarxx to be ported to different platforms.
For example mac, qnx or windows

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
character-devices: add support for char special devices

this commit fixes some issues regarding permissions
and adds propper support for character devices

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
streaming: add fifo and char special devices

add support for fifo and char special devices
to allow adding these files via streaming

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
streaming-api: improve safety

improve safety of streaming api by making sure no headers
are written before the file is finished

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
lz4: improve testing for lz4 api

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
format: re-format code

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
hard-links: add support for hard links

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
ustar: add proper directory support for ustar

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
copyright: update year

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
tests: add a test to make sure sockets can't be added

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
relative-paths: improve handling of relative paths

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
ci: fix build

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

thilo-schmitt commented on include/tarxx.h:
```diff
@@ -172,17 +172,51 @@ namespace tarxx {
+                lz4_flush();
             }
+#    endif
+            block_t block;
```
unused - please remove

alexmohr replied:
it's not present anymore in the latest version. at least I can't find it. It seems you commented on an older commit?

thilo-schmitt replied:
hmm... was in commit d283cbff12bed7f9840237868c6ca92528aef94c
interactive rebasing is your friend.
but not going to insist on it.

alexmohr replied:
fixed

thilo-schmitt commented on include/tarxx.h:
```diff
@@ -273,6 +276,16 @@ namespace tarxx {
             return std::filesystem::exists(path);
         };
+
+        [[nodiscard]] std::pair<bool, std::string> file_equivalent_present(const std::string& path, const std::unordered_set<std::string>& stored_files) const override
```
you might want to consider using `std::optional<std::string>` isntead of `std::pair<bool, std::string>`

alexmohr replied:
fixed

thilo-schmitt commented:
I have a feeling that code formatting at the end of your commit sequence is not in project style. can you please check and, if necessary, add another code reformatting commit at the end?

alexmohr commented:
> I have a feeling that code formatting at the end of your commit sequence is not in project style. can you please check and, if necessary, add another code reformatting commit at the end?

double checked again with clang-format-12. It's properly formatted already.
Either clang-format-10 nor clang-format-11 support all options set by the `.clang-format` file. 
As clang-format-13 isn't available for ubuntu 20.04 it's likely that it wasn't used for any of the prior commits. But I'm not entirely sure which clang format version I used.

thilo-schmitt commented:
> > I have a feeling that code formatting at the end of your commit sequence is not in project style. can you please check and, if necessary, add another code reformatting commit at the end?
> 
> double checked again with clang-format-12. It's properly formatted already. Either clang-format-10 nor clang-format-11 support all options set by the `.clang-format` file. As clang-format-13 isn't available for ubuntu 20.04 it's likely that it wasn't used for any of the prior commits. But I'm not entirely sure which clang format version I used.

sounds good to me. thanks for double checking, appreciate it!

thilo-schmitt commented:
in commit
[character-devices: add a test for char special devices](https://github.com/thilo-schmitt/tarxx/pull/16/commits/8428c2aeab0b8135dbfec1f35194b6b5571f566b)
and also
[fifo: add tests for fifo](https://github.com/thilo-schmitt/tarxx/pull/16/commits/466ecdd5668ba70d5a0a182e7ca2d727c9eef968)
it appears you are not only adding test cases, but also the capability itself.
would you mind making this more clear in your commit message?

thilo-schmitt commented:
all in all looks great. despite some formatting seesaw and some nitpicks, I don't see any real problems. please have a look on the comments and if we find some agreement there, the PR seems good to go.

alexmohr commented:
> in commit [character-devices: add a test for char special devices](https://github.com/thilo-schmitt/tarxx/pull/16/commits/8428c2aeab0b8135dbfec1f35194b6b5571f566b) and also [fifo: add tests for fifo](https://github.com/thilo-schmitt/tarxx/pull/16/commits/466ecdd5668ba70d5a0a182e7ca2d727c9eef968) it appears you are not only adding test cases, but also the capability itself. would you mind making this more clear in your commit message?

Fixed

alexmohr commented:
> all in all looks great. despite some formatting seesaw and some nitpicks, I don't see any real problems. please have a look on the comments and if we find some agreement there, the PR seems good to go.

Did you forget to add the comments? I'm not seeing anymore comments expect the formatting thing we discussed earlier and the commit message content.

alexmohr committed:
ci: add missing cmake flag to enable tests in runner

cmake was missing the WITH_TESTS flag and thus
tests were disabled in pipeline

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

thilo-schmitt commented:
> 



> > all in all looks great. despite some formatting seesaw and some nitpicks, I don't see any real problems. please have a look on the comments and if we find some agreement there, the PR seems good to go.
> 
> Did you forget to add the comments? I'm not seeing anymore comments expect the formatting thing we discussed earlier and the commit message content.

Uh-oh. Didn't submit the review :-) Sorry about that. It's only two comments anyway ...

thilo-schmitt commented:
Oh, snap, Docker gives you a little attitude ... appreciate if you'd look into a solution for this.

alexmohr committed:
tests: fix broken unit tests

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

thilo-schmitt commented on include/tarxx.h:
```diff
@@ -424,6 +491,54 @@ namespace tarxx {
+            if (name.size() <= UNIX_V7_USTAR_HEADER_LEN_NAME || type_ == tar_type::unix_v7) {
+                write_name_unix_v7_format();
+            } else {
+                const auto last_slash_index = name.rfind('/');
```
separating at the "**last**" "**slash**" - two words, two problems :-)
1. on other platforms the directory separator might be different, e.g. Windows uses a backslash (`\`) - this probably needs to go into the OS abstraction layer.
2. tar doesn't mandate separating at the _last_ slash, but at _some_ slash (or more precisely: at _a_ slash only). there could be cases where would have to separate the whole path+filename because it doesn't fit in the v7 field, but using the last slash just won't do the trick. like so `insanely_long_path_name_in_front_.../and_yet_another_very_long_path_component_.../my.file` ...

alexmohr replied:
take a look at the last commit, should be fixed now

alexmohr commented:
> Oh, snap, Docker gives you a little attitude ... appreciate if you'd look into a solution for this.

It's not docker but the tests are actually broken. I didn't notice this when testing locally because group name and user name was the same. In the runner it differs. It's fixed now. It was only an error in the tests

alexmohr committed:
file_equivalent_present: replace std::pair with std::optional

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>

alexmohr committed:
ustar: fix path separation

this commit changes how paths are split up

Signed-off-by: Alexander Mohr <alexander.m.mohr@mercedes-benz.com>
